### DISCUSSION TITLE
how to preserve YAML anchor ?

### DISCUSSION THREAD

**Discussion by sourcevault** [eemeli/yaml#303] [Q&A]
<img width="846" alt="preserve_anchors" src="https://user-images.githubusercontent.com/20347647/130606107-1bef8aed-70a8-4efc-a493-025ea18584ed.png">

Is there a way to preserve YAML anchor ? when using `.stringify` ? after making some changes using `yaml.parseDocument`.

**Comment by sourcevault** [comment]
I found the answer. 

You **should not** use `yaml.stringify` but use JS's `String` function, to preserve formatting, anchor name, etc.

In case somebody else hits the same problem.

Could somebody elaborate on the differences between `yaml@1` and `yaml@2` ? why did `yaml@2` get rid of `yaml/types` ? are there new patterns for creating nodes ?

**Reply by eemeli** [reply]
It would appear that your original issue was perhaps related to using the generic `YAML.stringify(any)` interface to stringify a YAML Document instance, yes? And indeed the right thing to do there is to use the document's own stringifier.

For the current interfaces, I would recommend looking at the [documentation site](https://eemeli.org/yaml/).

For details on changes made in the v2 work, I would recommend lookign at the [release notes](https://github.com/eemeli/yaml/releases).

**Reply by sourcevault** [reply]
Yes @eemeli thanks for the reply 🙂.

I wanted to close the discussion, since I found the answer, but it's a good idea to keep it in case somebody searches for "* preserve YAML anchor".

Is there a reason to prefer `.setIn` instead of doing the edits yourself using `new yamlTypes.Pair`,`new yamlTypes.Scalar`,......,etc ?

Are you exposing a len's type interface from functional programming to make things easier for user ?

**Reply by eemeli** [reply]
Yeah, the library isn't strongly opinionated about exactly how exactly it should be used. For most tasks there's more than one way of getting it done, so you can go with whatever suits your needs or desires. `setIn` specifically does a few more things automatically, such as creating any missing intermediary maps & sequences if they're not yet present in your document.

Feel free to mark your question as answered; that's pretty much the closest GH has for "closing" a discussion. It should still show up in search results.

**Reply by mwaeckerlin** [reply]
`document.toString()` and `String(document)` don't work for me, this code (file is a `docker-compose.yaml`):
```
const file = process.argv[2]
const project = process.argv[3]
const fs = require('fs')
const yaml = require('yaml')

const compose = yaml.parse(fs.readFileSync(file, 'utf-8'))

fs.writeFileSync(file, compose.toString())
```

Creates a file containing only:
```
[object Object]
```

Using:
```
    "yaml": "2.4.2"
```

It works with `yaml.stringify(compose)`.

**Comment by JuanVqz** [comment]
Hi! I'm using this package on [alacritty-themes](https://github.com/rajasegar/alacritty-themes) repository and we have the same problem I spend some time looking at the documentation, issues, prs but nothing did the trick, then @sourcevault you are right `String` did the work, I was looking for some function that can merge two string yamls or something like that, 🙏 thank you! 

BTW, you always are welcome to follow us on [**alacritty-themes**](https://github.com/rajasegar/alacritty-themes)😆

**Comment by thomasqbrady** [comment]
@eemeli When I use String() or toString() the YAML comes out in a format that I can't seem to change—curly braces and double quotation marks, resembling JSON, like this:
```
{
  "type": "Colors, Alphas, Gradients",
  "documentation":
    {
      "general": "Butter Colors. Values within the `base` name space are final, and
        not able to be overridden by applications implementing Butter.",
      "ios": "Specific documentation for iOS goes here. Usage, etc.",
      "android": "Specific documentation for Android goes here. Usage, etc.",
      "web": "Specific documentation for web goes here."
    },
  "tokens":
    {
```
Whereas the YAML.stringify function returns something that looks like this:
```
type: Colors, Alphas, Gradients
documentation:
  general: Butter Colors. Values within the `base` name space are final, and not able to be overridden by applications implementing Butter.
  ios: Specific documentation for iOS goes here. Usage, etc.
  android: Specific documentation for Android goes here. Usage, etc.
  web: Specific documentation for web goes here.
tokens:
```
Is there a configuration I'm missing (I've tried simpleKeys and versions 1.0-2) to get that format from toString (so I can keep the anchors)?

**Reply by eemeli** [reply]
Are you sure you're calling the document's `toString()` method, and not that of its contents? Proper YAML output is only provided for the document; when stringifying individual nodes an internally-maintained context object is required, otherwise there's a fallback to JSON output.

**Reply by thomasqbrady** [reply]
I'm creating the document with a call to `YAML.parseDocument()`, to which I'm passing a JSON.stringify'ed Javascript object.

I'm then doing transformations on that object via its contents/nodes. Then, finally, on that same object I got back from `YAML.parseDocument()` I'm calling `.toString()`.

**Reply by eemeli** [reply]
Ah, okay. If you're creating a YAML document from a JSON-stringified source, then it's a feature that the "double quoting" and flow rather than block collection styles are retained. If you have a JS value, you can use `new YAML.Document(value)` to create the YAML doc. See documentation here: https://eemeli.org/yaml/#creating-documents

**Reply by thomasqbrady** [reply]
When I pass my JS object to new YAML.Document, a la `let yamlTokens = new YAML.Document(mergedTokens)`, where `mergedTokens` is a JS object, I get a YAML document whose `contents` is null. When I create the document the way I was before (passing a JSON.stringify'ed object to YAML.parseDocument), I'm getting a working YAML doc with contents as I expect...

**Reply by thomasqbrady** [reply]
Hmm. Looking at the document I'm getting back from new YAML.Document(mergedTokens), it looks like the mergedTokens object is getting turned into the options property of the new document, rather than the contents...

**Reply by thomasqbrady** [reply]
I see. Your link went to version 2 docs. I'm using version 1 at the moment, so the Document constructor expects only an options object. In version one it looks like I can do this safely?
```
          yamlTokens = new YAML.Document()
          yamlTokens.contents = mergedTokens
```

**Reply by thomasqbrady** [reply]
No. Heh. That results in a `contents` property that is not a collection of Node/Map/etc.

**Comment by thomasqbrady** [comment]
I upgraded to 2.0 and it seems to be working now, although I can't figure out how to find the list of anchors in 2.0. In 1.10.0 the Document had an anchors property. In the documentation for 2.0 the "working with anchors" section linked in the table of contents is missing… is 2.0 a beta?

**Reply by eemeli** [reply]
Yes, `yaml@2` is still in prerelease; "beta" if you will. And yes, the TOC link in the docs appears to have been broken when the anchor API was refactored in #248. Should fix that. Anchors are now mentioned here: https://eemeli.org/yaml/#content-nodes

If you need to do larger operations on anchors, I'd recommend using a visitor: https://eemeli.org/yaml/#modifying-nodes

**Reply by thomasqbrady** [reply]
Thanks! Will check I out.
